lib/rubocop/cop/rails/active_model_errors_direct_manipulation.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # This cop checks direct manipulation of ActiveModel#errors as hash.
+      # These operations are deprecated in Rails 6.1 and will not work in Rails 7.
+      #
+      # @example
+      #   # bad
+      #   user.errors[:name] << 'msg'
+      #   user.errors.messages[:name] << 'msg'
+      #
+      #   # good
+      #   user.errors.add(:name, 'msg')
+      #
+      #   # bad
+      #   user.errors[:name].clear
+      #   user.errors.messages[:name].clear
+      #
+      #   # good
+      #   user.errors.delete(:name)
+      #
+      class ActiveModelErrorsDirectManipulation < Base
+        MSG = 'Avoid manipulating ActiveModel errors as hash directly.'
+
+        MANIPULATIVE_METHODS = ':<< :append :clear :collect! :compact! :concat '\
+                               ':delete :delete_at :delete_if :drop :drop_while :fill :filter! :keep_if '\
+                               ':flatten! :insert :map! :pop :prepend :push :reject! :replace :reverse! '\
+                               ':rotate! :select! :shift :shuffle! :slice! :sort! :sort_by! :uniq! :unshift'
+
+        BASE_NODE = {
+          general: '{send ivar lvar}',
+          model_file: '{nil? send ivar lvar}'
+        }.freeze
+
+        PATTERN = {
```

Comment:
Would you agree to separate those patterns and define them statically? It's possible to combine them with `{ }`.
Like:
```ruby
MANIPULATIVE_METHODS = Set[:<<, :append, ...].freeze

def_node_pattern :root_manipulation?, <<~PATTERN
  (send
    (send
      (send %1 :errors) :[] ...)
      MANIPULATIVE_METHODS
      ...
    )
  )
PATTERN

def_node_pattern :any_manipulation?, <<~PATTERN
  {
    root_manipulation?(%1)
    root_assignment?(%1)
  }
PATTERN

def_node_pattern :general?, <<~PATTERN
  {send ivar lvar}
PATTERN

def on_send(node)
  send_pattern = 
    '{nil? send ivar lvar}' # I might be mistaken, check docs https://docs.rubocop.org/rubocop-ast/node_pattern.html#composing-complex-expressions-with-multiple-matchers
  else
    :general? # or like this? https://github.com/rubocop/rubocop-rspec/blob/573f22bbbd54e5332231b9823b87843550ae21a4/lib/rubocop/cop/rspec/multiple_expectations.rb#L91
  end

  any_manipulation?(node, send_pattern) do |node|
    add_offence(...)
  end
end
```

Reply:
Sure if we can avoid all these string interpolation. However I actually know very little about how the rubocop-ast works. I tried to apply what you provided, but it would say

> undefined method `def_node_pattern' for RuboCop::Cop::Rails::ActiveModelErrorsDirectManipulation:Class
> Did you mean?  def_node_matcher

When I tried using `def_node_matcher` it would say:

> RuboCop::AST::NodePattern::Invalid:
> parse error on value :")" (")")

If you could pull some AST expertise on this it would be greatly appreciated.

Reply:
Let's try it step by step.
Constant simplification:
![image](https://user-images.githubusercontent.com/6916/147595900-c2d82b9c-0ece-4eaa-8e1d-1efded676094.png)

Reply:
I'll just send the whole thing as I see a canonical implementation.
Leaving it up to you if you find it readable - it's up to you to support the cop (hopefully :) anyway.

```ruby
      class ActiveModelErrorsDirectManipulation < Base
        MSG = 'Avoid manipulating ActiveModel errors as hash directly.'

        MANIPULATIVE_METHODS = Set[%i[<< append clear collect! compact! concat
                               delete delete_at delete_if drop drop_while fill filter! keep_if
                               flatten! insert map! pop prepend push reject! replace reverse!
                               rotate! select! shift shuffle! slice! sort! sort_by! uniq! unshift]].freeze

        def_node_matcher :general?, '{send ivar lvar}'
        def_node_matcher :model?, '{nil? send ivar lvar}'

        def_node_matcher :any_manipulation?, <<~PATTERN
          {
            #root_manipulation?
            #root_assignment?
            #messages_details_manipulation?
            #messages_details_assignment?
          }
        PATTERN

        def_node_matcher :root_manipulation?, <<~PATTERN
          (send
            (send
              (send #file_type_match? :errors) :[] ...)
            #MANIPULATIVE_METHODS
            ...
          )
        PATTERN

        def_node_matcher :root_assignment?, <<~PATTERN
          (send
            (send #file_type_match? :errors)
            :[]=
            ...)
        PATTERN

        def_node_matcher :messages_details_manipulation?, <<~PATTERN
          (send
            (send
              (send
                (send #file_type_match? :errors)
                {:messages :details})
                :[]
                ...)
              #MANIPULATIVE_METHODS
            ...)
        PATTERN

        def_node_matcher :messages_details_assignment?, <<~PATTERN
          (send
            (send
              (send #file_type_match? :errors)
              {:messages :details})
            :[]=
            ...)
        PATTERN

        def on_send(node)
          any_manipulation?(node) do
            add_offense(node)
          end
        end

        private

        def file_type_match?(node)
          model_file? ? model?(node) : general?(node)
        end

        def model_file?
          processed_source.buffer.name.include?('/models/')
        end
      end
```

Reply:
Thanks so much! I made some renamings and it works. I've placed this as a separate commit with you being the author.